### PR TITLE
Fix withCredetials() to correctly detect missing creds

### LINKED ISSUE #1155: Problematic regression with ~/.github file authentication
Authenticating with `~/.github` when using `GitHub.connect()` doesn't work anymore in recent versions of the GitHub API.

The client "authenticates" as anonymous instead of using the credentials.

The issue is that the `GitHubBuilder.authorizationProvider` field is never null as it's now initialized to an `ANONYMOUS` instance here: https://github.com/hub4j/github-api/blob/main/src/main/java/org/kohsuke/github/GitHubBuilder.java#L35 .

Thus when arriving here https://github.com/hub4j/github-api/blob/main/src/main/java/org/kohsuke/github/GitHubBuilder.java#L63 , the field is never null and we never get the chance to use the  `~/.github` property file to authenticate a few lines below.

I'm not exactly sure how we want to fix this. A comment in `GitHubBuilder.fromCredentials()` says `@throws IOException If there are no credentials defined in the ~/.github properties file or the process environment.` which is obviously not true anymore.

So:
- we definitely need to test that `authorizationProvider` is not `ANONYMOUS` here: https://github.com/hub4j/github-api/blob/main/src/main/java/org/kohsuke/github/GitHubBuilder.java#L35 .
- I'm not sure if we want to fix the comment and allow anonymous access (but given we explicitly called `fromCredentials()`, I would rather throw an exception) or throw an exception again if we don't have any credentials. The latter would obviously prevent anonymous access.

Note that it can be a bit hard to debug as in my case the GH API was just throwing 404 Not Found errors when trying to do some write operations instead of complaining about the missing credentials.

Also, no idea on how to test this :).

Thoughts?

### PR DESCRIPTION
**Pull request by bitwiseman**
# Description

Fixes #1155

@gsmet @hcoles

If you want to test this fix manually against #1156 or #1157 , that would be helpful. 

# Before submitting a PR:
We love getting PRs, but we hate asking people for the same basic changes every time.

- [x] Push your changes to a branch other than `main`. Create your PR from that branch.
- [x] Add JavaDocs and other comments
- [x] Write tests that run and pass in CI. See [CONTRIBUTING.md](CONTRIBUTING.md) for details on how to capture snapshot data.
- [x] Run `mvn -D enable-ci clean install site` locally. If this command doesn't succeed, your change will not pass CI.

# When creating a PR: 

- [x] Fill in the "Description" above.
- [x] Enable "Allow edits from maintainers".

### CODE REVIEW

File: src/test/java/org/kohsuke/github/GitHubConnectionTest.java [hunk 1]
```diff
@@ -137,6 +138,67 @@ public void testGitHubBuilderFromCustomEnvironment() throws IOException {
...
+        assertThat(builder.authorizationProvider.getEncodedAuthorization(), equalTo("Bearer bogus jwt token string"));
+
+        props.put("password", "bogus weak password");
+        setupEnvironment(props);
+        builder = GitHubBuilder.fromCredentials();
+
+        assertThat(builder.authorizationProvider, instanceOf(UserAuthorizationProvider.class));
+        assertThat(builder.authorizationProvider.getEncodedAuthorization(),
+                equalTo("Basic Ym9ndXMgbG9naW46Ym9ndXMgd2VhayBwYXNzd29yZA=="));
+        assertThat(((UserAuthorizationProvider) builder.authorizationProvider).getLogin(), equalTo("bogus login"));
```
Comment (bitwiseman) [inline src/test/java/org/kohsuke/github/GitHubConnectionTest.java:181]: TODO: We really should also test for the `~/.github` scenario as well.

**Comment by hcoles** [comment]
@bitwiseman I'm afriad this does not resolve #1157

**Review by gsmet — APPROVED** [review]
Perfect, thanks!

**Comment by gsmet** [comment]
Yeah #1157 looks like another issue.

**Pull request merged**